tower-util/src/ext/mod.rs

```diff
@@ -0,0 +1,98 @@
+//! Combinators for working with `Service`s
+
+use futures::{Future, IntoFuture};
+use tower_service::Service;
+
+mod and_then;
+mod apply;
+mod map;
+mod map_err;
+mod then;
+
+pub use self::and_then::AndThen;
+pub use self::apply::Apply;
+pub use self::map::Map;
+pub use self::map_err::MapErr;
+pub use self::then::Then;
+
+impl<T: ?Sized> ServiceExt for T where T: Service {}
+
+/// An extension trait for `Service`s that provides a variety of convenient
+/// adapters
+pub trait ServiceExt: Service {
+    fn apply<F, R, Req>(self, f: F) -> Apply<Self, F, R, Req>
+    where
+        Self: Clone + Sized,
+        Self::Error: Into<<R::Future as Future>::Error>,
+        F: Fn(Req, Self) -> R,
+        R: IntoFuture,
+    {
+        Apply::new(f, self)
+    }
+
+    /// Call another service after call to this one has resolved successfully.
+    ///
+    /// This function can be used to chain two services together and ensure that
+    /// the second service isn't called until call to the fist service have finished.
+    /// Result of the call to the first service is used as an input parameter
+    /// for the second service's call.
+    ///
+    /// Note that this function consumes the receiving service and returns a
+    /// wrapped version of it.
+    fn and_then<B, U>(self, service: U) -> AndThen<Self, B>
+    where
+        Self: Sized,
+        U: Into<B>,
```

Comment:
this would be useful for #69, but i found that `From/Into` does not work well with type inference.
we need to use custom `IntoService` trait. i can remove conversion or implement IntoService, it is up to you?

Reply:
Yeah, I think we should limit `Into` / `From` bounds due to type inference issues.

Let's remove the conversion for now.

Separately, I think `IntoService` will helpful at some point, but lets punt for this PR so that we can get it merged.